### PR TITLE
Add list knowledge endpoint

### PR DESCRIPTION
**Pull request by alistairknight**
# **What this PR does / why we need it**:
*A brief description of changes being made*
*Link to the relevant work items: e.g: Relates to ADO Work Item AB#213700 and builds on #3376 (link to ADO Build ID URL)*

# **Special notes for your reviewer**
*Any specific actions or notes on review?*

# Testing
*Any relevant testing information and pipeline runs.*

# Checklist (please delete before completing or setting auto-complete)
- [ ] Story Work items associated (not Tasks)
- [ ] Successful testing run(s) link provided
- [ ] Title pattern should be `{work item number}: {title}`
- [ ] Description covers all the changes in the PR
- [ ] This PR contains documentation
- [ ] This PR contains tests
- [ ] Version number in **package.json** has been updated

# **How does this PR make you feel**:
gif

### CODE REVIEW

File: app/routes/knowledge.js [hunk 1]
```diff
@@ -1,9 +1,20 @@
 const Joi = require('joi')
 
 const { processPayloadFile } = require('../lib/file')
-const { saveKnowledge, updateKnowledgeMetadata } = require('../storage/knowledge-document-repo')
+const { listKnowledge, saveKnowledge, updateKnowledgeMetadata } = require('../storage/knowledge-document-repo')
 
 module.exports = [{
+  method: 'GET',
+  path: '/knowledge',
+  options: {
```
Comment (steve-dickinson) [inline app/routes/knowledge.js:9]: Could you put a validation check on the the request query and handle the error.

File: app/routes/knowledge.js [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
+  },
+  handler: async (request, h) => {
+    const { category } = request.query
+    const knowledge = await listKnowledge(category)
+    return h.response(knowledge).code(201)
```
Comment (steve-dickinson) [inline app/routes/knowledge.js]: This needs to be a 200 and not a 201 as it isn't something which has been created.

File: app/storage/knowledge-document-repo.js [hunk 3]
```diff
@@ -5,6 +5,45 @@ const { mapMetadataToBlob, mapMetadataToBase } = require('../mappers/knowledge-m
...
+  for await (const blob of knowledgeContainer.listBlobsFlat(listOptions)) {
+    const metadata = mapMetadataToBase(blob.metadata)
+
+    blob.metadata = metadata
+    if (category === '' || metadata.category === category) {
+      blobs.push(blob)
+    }
+  }
+
+  blobs.sort((a, b) => {
```
Comment (steve-dickinson) [inline app/storage/knowledge-document-repo.js]: Could maybe make this function a bit cleaner by moving out the sort into its own function.

**Review by steve-dickinson — CHANGES_REQUESTED** [review]
Just a few changes and suggestions

File: app/routes/knowledge.js (hunk 2 above)
Comment (alistairknight) [reply app/routes/knowledge.js]: Sorted

File: app/routes/knowledge.js (hunk 1 above)
Comment (alistairknight) [reply app/routes/knowledge.js:9]: Validation added

File: app/storage/knowledge-document-repo.js (hunk 3 above)
Comment (alistairknight) [reply app/storage/knowledge-document-repo.js]: Refactored

**Review by steve-dickinson — DISMISSED** [review]
LGTM

**Review by sfitz42 — APPROVED** [review]
LGTM

Might need a warning though if there is no metadata

**Pull request merged**